src/expr/parser.rs

```diff
@@ -25,34 +29,49 @@ named!(i64_literal<i64>,
   )
 );
 
-named!(factor<i64>,
+named!(expr_literal<Expr>,
+  map!(
+    isize_literal,
+    Expr::Literal
+  )
+);
+
+named!(factor<Expr>,
   alt!(
-    i64_literal
+    expr_literal
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
              many0!(
                alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
+                 map!(preceded!(tag!("*"), factor),
+                    |mul| { let new_acc = mem::replace(&mut acc, Expr::Literal(0));
+                            acc = Expr::BinOp(Box::new(new_acc), BinOp::Times, Box::new(mul))}) |
+                 map!(preceded!(tag!("/"), factor),
+                    |div| { let new_acc = mem::replace(&mut acc, Expr::Literal(0));
+                            acc = Expr::BinOp(Box::new(new_acc), BinOp::Over, Box::new(div))})
                )
              ),
     || { return acc }
   )
 );
 
-named!(pub expr <i64>,
+named!(pub expr <Expr>,
   chain!(
     mut acc: term  ~
              many0!(
                alt!(
-                 tap!(add: preceded!(tag!("+"), term) => acc = acc + add) |
-                 tap!(sub: preceded!(tag!("-"), term) => acc = acc - sub)
+                 map!(preceded!(tag!("+"), term),
+                  |add| { let new_acc = mem::replace(&mut acc, Expr::Literal(0));
+                            acc = Expr::BinOp(Box::new(new_acc), BinOp::Plus, Box::new(add))})|
+                 map!(preceded!(tag!("-"), term),
+                  |sub| { let new_acc = mem::replace(&mut acc, Expr::Literal(0));
+                            acc = Expr::BinOp(Box::new(new_acc), BinOp::Minus, Box::new(sub))})
```

Comment:
You us this pattern frequently -- it would have been a good thing to abstract out.